src/poc/miner_poc_mgr.erl

```diff
@@ -0,0 +1,1001 @@
+    #state{chain = Chain} = State
+)->
+    Ledger = blockchain:ledger(Chain),
+    CurPOCChallengerType =
+        case blockchain:config(?poc_challenger_type, Ledger) of
+            {ok, V}  -> V;
+            _ -> undefined
+        end,
+    lager:debug("received poc keys event, sync is ~p, poc_challenge_type is ~p", [_Sync, CurPOCChallengerType]),
+    State1 = maybe_init_addr_hash(State),
+    ok = process_block_pocs(CurPOCChallengerType, BlockHeight, BlockHash, BlockPOCs, Chain, State1),
+    {noreply, State1};
+handle_info(_Info, State = #state{}) ->
+    {noreply, State}.
+
+terminate(_Reason, _State = #state{}) ->
+    persistent_term:erase(?POC_DB_CF),
+    ok.
+
+%%%===================================================================
+%%% breakout functions
+%%%===================================================================
+-spec handle_add_block_event(
+    POCChallengeType :: validator | undefined,
+    BlockHash :: binary(),
+    Chain :: blockchain:blockchain(),
+    State :: state()
+) -> ok.
+handle_add_block_event(POCChallengeType, BlockHash, Chain, State) when POCChallengeType == validator ->
+    case blockchain:get_block(BlockHash, Chain) of
+        {ok, Block} ->
+            BlockHeight = blockchain_block:height(Block),
+            %% take care of GC
+            ok = purge_local_pocs(Block, State),
+            Ledger = blockchain:ledger(Chain),
+
+            %% GC local pocs keys every 50 blocks
+            case BlockHeight rem 50 == 0 of
+                true ->
+                    ok = purge_local_poc_keys(BlockHeight, Ledger);
+                false ->
+                    ok
+            end;
+        _ ->
+            %% err what?
+            ok
+    end;
+handle_add_block_event(_POCChallengeType, _BlockHash, _Chain, _State) ->
+    ok.
+
+-spec handle_witness(
+    Witness :: blockchain_poc_witness_v1:poc_witness(),
+    OnionKeyHash :: binary(),
+    Address :: libp2p_crypto:pubkey_bin(),
+    State :: #state{}
+) -> {noreply, state()}.
+handle_witness(Witness, OnionKeyHash, Peer, #state{chain = Chain} = State) ->
+    lager:debug("got witness ~p with onionkeyhash ~p", [Witness, OnionKeyHash]),
+    GatewayWitness = blockchain_poc_witness_v1:gateway(Witness),
+    case miner_poc_denylist:check(GatewayWitness) of
+        true ->
+            lager:notice("dropping witness from ~p due to denylist", [libp2p_crypto:bin_to_b58(GatewayWitness)]),
+            {noreply, State};
+        false ->
+            %% Validate the witness is correct
+            Ledger = blockchain:ledger(Chain),
+            case validate_witness(Witness, Ledger) of
+                false ->
+                    lager:warning("ignoring witness ~p for onionkeyhash ~p. Reason: invalid", [Witness, OnionKeyHash]),
+                    {noreply, State};
+                true ->
+                    %% get the local POC
+                    case ?MODULE:local_poc(OnionKeyHash) of
```

Comment:
Might be better to check this is a local poc before bothering to call validate_witness?

Reply:
done

Reply:
https://github.com/helium/miner/pull/1068/commits/32247902ed59dcdcf354746ce707c1a1aa4c5eda